hugr-llvm/src/sum/layout.rs

```diff
@@ -0,0 +1,216 @@
+//! The algorithm for computing the fields of the struct representing a
+//! [HugrSumType].
+//!
+use std::borrow::Cow;
+use std::cmp::Reverse;
+use std::collections::BTreeMap;
+use std::fmt;
+use std::ops::RangeFrom;
+
+use inkwell::types::{BasicType, BasicTypeEnum};
+use itertools::Itertools as _;
+
+use super::elidable_type;
+
+/// Compute the layout of the non-tag fields of the struct representing a
+/// HugrSumType
+///
+/// The first return value are the non-tag fields of the struct.
+///
+/// The second return value is a per-variant-per-field `Option<usize>`. If that
+/// value is `None`, then the field should be elided from the struct.  Values
+/// for this field should be materialized via `undef`. If that value is
+/// `Some(i)` then values of this field should be read and written to the `i`th
+/// field of the struct.
+pub(super) fn layout_variants<'c>(
+    variants: impl AsRef<[Vec<BasicTypeEnum<'c>>]>,
+) -> (Vec<BasicTypeEnum<'c>>, Vec<Vec<Option<usize>>>) {
+    let variants = variants.as_ref();
+    let (sorted_fields, layout) = layout_variants_impl::<BasicTypeOrd>(
+        variants
+            .iter()
+            .cloned()
+            .map(|x| x.into_iter().map_into().collect_vec())
+            .collect_vec(),
+        |x| elidable_type(x.0),
+    );
+    let sorted_fields = sorted_fields.into_iter().map_into().collect_vec();
+    (sorted_fields, layout)
+}
+
+fn size_of_type<'c>(t: impl BasicType<'c>) -> Option<u64> {
+    match t.as_basic_type_enum() {
+        BasicTypeEnum::ArrayType(t) => t.size_of().and_then(|x| x.get_zero_extended_constant()),
+        BasicTypeEnum::FloatType(t) => t.size_of().get_zero_extended_constant(),
+        BasicTypeEnum::IntType(t) => t.size_of().get_zero_extended_constant(),
+        BasicTypeEnum::PointerType(t) => t.size_of().get_zero_extended_constant(),
+        BasicTypeEnum::StructType(t) => t.size_of().and_then(|x| x.get_zero_extended_constant()),
+        BasicTypeEnum::VectorType(t) => t.size_of().and_then(|x| x.get_zero_extended_constant()),
+    }
+}
+
+#[derive(derive_more::Debug, Clone, PartialEq, Eq)]
+/// `BasicTypeEnum` does not `impl` `Ord`. We use this type to put an ordering `BasicTypeEnum`,
+/// first by reverse-size and then by, string representation.
+struct BasicTypeOrd<'c>(
+    BasicTypeEnum<'c>,
+    #[debug(skip)] u64,
+    #[debug(skip)] Cow<'c, String>,
+);
```

Comment:
Not a big fan of the string ordering performance wise, but I can't think of a better solution either that doesn't require a lot of work...